client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -1155,23 +1013,199 @@ private int getCloseConnectionTimeout() {
     }
 
     /**
-     * Wrapped connection handler used in the {@link #connect()} method. Allows adding an additional handler.
+     * Encapsulates a connection attempt (with possible retries).
      */
-    private static class ConnectMethodConnectionHandler implements Handler<AsyncResult<HonoConnection>> {
+    private class ConnectionAttempt {
 
+        private final ProtonClientOptions clientOptions;
         private final Handler<AsyncResult<HonoConnection>> connectionHandler;
-        private final Handler<AsyncResult<HonoConnection>> additionalHandler;
+        private final AtomicInteger connectAttempts = new AtomicInteger(0);
+        private final AtomicBoolean cancelled = new AtomicBoolean();
+
+        private Long reconnectTimerId;
+
+        ConnectionAttempt(final ProtonClientOptions clientOptions, final Handler<AsyncResult<HonoConnection>> connectionHandler) {
+            this.clientOptions = clientOptions;
+
+            this.connectionHandler = ar -> {
+                if (ar.failed()) {
+                    clearState();
+                }
+                currentConnectionAttempt.compareAndSet(this, null);
+                connectionHandler.handle(ar);
+                deferredConnectionCheckHandler.setConnectionAttemptFinished(ar);
```

Comment:
IMHO this could/should be done in HonoConnectionImpl.connect, couldn't it?

Reply:
Same argument as above: I prefer to make all updates to both `currentConnectionAttempt` and `deferredConnectionCheckHandler` close to one another for better readability (the other place is `ConnectionAttempt.start()` then).

Reply:
I see